code/ship/ship.cpp

```diff
@@ -11897,92 +11895,35 @@ void ship_process_targeting_lasers()
- * 
- *	Calls ::weapon_hit(), indirectly via ::weapon_detonate(), to detonate weapon.
- *	If it's a weapon that spawns particles, those will be released.
- */
-static int maybe_detonate_weapon(ship_weapon *swp, object* /*src*/)
-{
-	int			objnum = swp->last_fired_weapon_index;
-	object		*objp;
-	weapon_info	*wip;
-
-	if ((objnum < 0) || (objnum >= MAX_OBJECTS)) {
-		return 0;
-	}
-	
-	objp = &Objects[objnum];
-
-	if (objp->type != OBJ_WEAPON){
-		return 0;
-	}
-
-	if ((objp->instance < 0) || (objp->instance >= MAX_WEAPONS)){
-		return 0;
-	}
-
-	// check to make sure that the weapon to detonate still exists
-	if ( swp->last_fired_weapon_signature != objp->signature ){
-		return 0;
-	}
-
-	Assert(Weapons[objp->instance].weapon_info_index != -1);
-	wip = &Weapon_info[Weapons[objp->instance].weapon_info_index];
-
-	if (wip->wi_flags[Weapon::Info_Flags::Remote]) {
-
-		int	weapon_sig;
-
-		weapon_sig = objp->signature;
-
-		if (swp->last_fired_weapon_signature == weapon_sig) {
-			weapon_detonate(objp);
-			swp->last_fired_weapon_index = -1;
-
-			return 1;
-		}
-	}
-
-	return 0;
-}
-
 /**
  * Maybe detonate secondary weapon that's already out.
  * @return Return true if we detonate it, false if not.
  */
-static int ship_fire_secondary_detonate(object *obj, ship_weapon *swp)
+static bool ship_fire_secondary_detonate(object *obj, ship_weapon *swp)
 {
-	if (swp->last_fired_weapon_index != -1)
+	if (swp->remote_detonaters_active > 0) {
 		if (timestamp_elapsed(swp->detonate_weapon_time)) {
-			object	*first_objp = &Objects[swp->last_fired_weapon_index];
-			if (maybe_detonate_weapon(swp, obj)) {
-				//	If dual fire was set, there could be another weapon to detonate.  Scan all weapons.
-				missile_obj	*mo;
-
-				// check for currently locked missiles (highest precedence)
-				for ( mo = GET_FIRST(&Missile_obj_list); mo != END_OF_LIST(&Missile_obj_list); mo = GET_NEXT(mo) ) {
-					object	*mobjp;
-					Assert(mo->objnum >= 0 && mo->objnum < MAX_OBJECTS);
-					mobjp = &Objects[mo->objnum];
-					if ((mobjp != first_objp) && (mobjp->parent_sig == obj->parent_sig)) {
-						if (Weapon_info[Weapons[mobjp->instance].weapon_info_index].wi_flags[Weapon::Info_Flags::Remote]) {
-							weapon_detonate(mobjp);
-						}
+			missile_obj	*mo;
+
+			// check for currently locked missiles (highest precedence)
+			for ( mo = GET_FIRST(&Missile_obj_list); mo != END_OF_LIST(&Missile_obj_list); mo = GET_NEXT(mo) ) {
+				object	*mobjp;
+				Assert(mo->objnum >= 0 && mo->objnum < MAX_OBJECTS);
+				mobjp = &Objects[mo->objnum];
+				if (mobjp->parent_sig == obj->parent_sig && Weapon_info[Weapons[mobjp->instance].weapon_info_index].wi_flags[Weapon::Info_Flags::Remote]) {
+					// dont detonate if this guy just spawned i.e. we just spawned him in a previous iteration of this loop
+					if (Missiontime - Weapons[mobjp->instance].creation_time > fl2f(0.01)) {
+						weapon_detonate(mobjp);
+						swp->remote_detonaters_active--;
 					}
```

Comment:
Why not check the `detonate_weapon_time` field here?

Even you leave the check as-is, I would still change it to 0.1 or 0.2 seconds.  A 120 FPS game could trigger twice within 0.01 seconds.

Reply:
As the comment says, this isn't catching possible retriggers from the player, that's caught above this with `detonate_weapon_time`. This is for not detonating weapons which have been created by this very loop in this frame because `weapon_detonate` is possibly going to create more `missile_objects`. Really, it should probably 0.

Reply:
Ah, I missed the `if (timestamp_elapsed(swp->detonate_weapon_time))` check at the top.

Ok, I guess this looks gold.